toolchain/check/convert.cpp

```diff
@@ -990,38 +991,49 @@ static auto PerformBuiltinConversion(Context& context, SemIR::LocId loc_id,
     }
   }
 
-  if (sem_ir.types().Is<SemIR::FacetType>(target.type_id)) {
-    auto lookup_inst_id = value_id;
-
-    // `FacetAccessType` wraps an instruction that evaluates to a facet value
-    // (of type `FacetType`). If the `FacetType` matches the target
-    // `FacetType` then we don't need to do impl lookup. Otherwise, we want to
-    // try convert the result of the instruction in the `FacetAccessType`.
-    if (auto facet_access_type_value =
-            context.insts().TryGetAs<SemIR::FacetAccessType>(lookup_inst_id)) {
-      auto facet_value_inst_id = facet_access_type_value->facet_value_inst_id;
-      if (context.insts().Get(facet_value_inst_id).type_id() ==
-          target.type_id) {
+  if (sem_ir.types().Is<SemIR::FacetType>(target.type_id) &&
+      (sem_ir.types().Is<SemIR::TypeType>(value_type_id) ||
+       sem_ir.types().Is<SemIR::FacetType>(value_type_id))) {
+    // The value is a type or facet value, so it has a constant value. We get
+    // that to unwrap things like NameRef and get to the underlying type or
+    // facet value instruction so that we can use `TryGetAs`.
+    auto lookup_inst_id = sem_ir.constant_values().GetConstantInstId(value_id);
+
+    if (auto facet_access_type_inst =
+            sem_ir.insts().TryGetAs<SemIR::FacetAccessType>(lookup_inst_id)) {
+      // Conversion from a `FacetAccessType` to a `FacetValue` of the target
+      // `FacetType` if the instruction in the `FacetAccessType` is of a
+      // `FacetType` that satisfies the requirements of the target `FacetType`.
+      auto facet_value_inst_id = facet_access_type_inst->facet_value_inst_id;
+
+      // If the `FacetType` exactly matches the target `FacetType` then we can
+      // shortcut and use that value, and avoid impl lookup.
+      if (sem_ir.insts().Get(facet_value_inst_id).type_id() == target.type_id) {
         return facet_value_inst_id;
       }
-      lookup_inst_id = facet_access_type_value->facet_value_inst_id;
+
+      auto witness_inst_id = LookupImplWitness(
+          context, loc_id, context.constant_values().Get(facet_value_inst_id),
+          context.types().GetConstantId(target.type_id));
+      if (witness_inst_id != SemIR::InstId::None) {
+        return context.AddInst<SemIR::FacetValue>(
+            loc_id, {
+                        .type_id = target.type_id,
+                        .type_inst_id = lookup_inst_id,
+                        .witness_inst_id = witness_inst_id,
+                    });
```

Comment:
```suggestion
            loc_id, {.type_id = target.type_id,
                     .type_inst_id = lookup_inst_id,
                     .witness_inst_id = witness_inst_id});
```

Minor nit, but since there aren't many members, you might consider dropping the suffix comma which would probably result in the above format. I _think_ we'll do suffix commas when there are a lot of members (like `Function`), but `AddInst` is usually using the shorter format.

Reply:
Realizing the stack here, will try to move this comment.

Reply:
Unresolving, this is apparently my one comment that actually *is* this PR.

Reply:
Done